### PR TITLE
Cross chain tokens esdt nft mint with same nonce

### PR DESCRIPTION
**Pull request by mariusmihaic**
- Allow nft esdt cross chain token mint with the same nonce(that it came from another sovereign chain)
- Nonce should be the last args in vm input and the address should be whitelisted
- Nonce will not be incremented and saved in account's storage

### CODE REVIEW

File: builtInFunctions/esdtNFTCreate.go [hunk 1]
```diff
@@ -116,16 +128,16 @@ func (e *esdtNFTCreate) ProcessBuiltinFunction(
...
 	}
-	lenArgs := len(vmInput.Arguments)
-	if lenArgs < minNumOfArgs {
+	argsLen := len(vmInput.Arguments)
+	if argsLen < minNumOfArgs {
 		return nil, fmt.Errorf("%w, wrong number of arguments", ErrInvalidArguments)
 	}
 
 	accountWithRoles := acntSnd
 	uris := vmInput.Arguments[6:]
```
Comment (axenteoctavian) [inline builtInFunctions/esdtNFTCreate.go:137]: if cross-chain mint then uris is `[6 : argsLen-1]`

File: builtInFunctions/esdtNFTCreate.go [hunk 2]
```diff
@@ -11,6 +11,7 @@ import (
 	"github.com/multiversx/mx-chain-core-go/data/esdt"
 	"github.com/multiversx/mx-chain-core-go/data/vm"
 	logger "github.com/multiversx/mx-chain-logger-go"
+
 	"github.com/multiversx/mx-chain-vm-common-go"
```
Comment (axenteoctavian) [inline builtInFunctions/esdtNFTCreate.go:15]: This import should be in the 2nd category

File: builtInFunctions/creator.go [hunk 3]
```diff
@@ -4,6 +4,7 @@ import (
 	"github.com/mitchellh/mapstructure"
 	"github.com/multiversx/mx-chain-core-go/core"
 	"github.com/multiversx/mx-chain-core-go/core/check"
+
 	vmcommon "github.com/multiversx/mx-chain-vm-common-go"
```
Comment (axenteoctavian) [inline builtInFunctions/creator.go:8]: This import should be in the 2nd category

File: builtInFunctions/esdtNFTCreate_test.go [hunk 4]
```diff
@@ -4,34 +4,42 @@ import (
...
 	"github.com/multiversx/mx-chain-core-go/core"
 	"github.com/multiversx/mx-chain-core-go/core/check"
 	"github.com/multiversx/mx-chain-core-go/data/esdt"
 	"github.com/multiversx/mx-chain-core-go/data/vm"
-	"github.com/multiversx/mx-chain-vm-common-go"
-	"github.com/multiversx/mx-chain-vm-common-go/mock"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+
+	"github.com/multiversx/mx-chain-vm-common-go"
```
Comment (axenteoctavian) [inline builtInFunctions/esdtNFTCreate_test.go:17]: These imports should be in the 2nd category

File: builtInFunctions/esdtNFTCreate.go [hunk 5]
```diff
@@ -19,61 +20,78 @@ var (
...
+	rolesHandler                  vmcommon.ESDTRoleHandler
+	funcGasCost                   uint64
+	gasConfig                     vmcommon.BaseOperationCost
+	esdtStorageHandler            vmcommon.ESDTNFTStorageHandler
+	enableEpochsHandler           vmcommon.EnableEpochsHandler
+	mutExecution                  sync.RWMutex
+	crossChainTokenCheckerHandler CrossChainTokenCheckerHandler
+}
+
+type ESDTNFTCreateFuncArgs struct {
```
Comment (sasurobert) [inline builtInFunctions/esdtNFTCreate.go:47]: comment is missing.

File: builtInFunctions/esdtNFTCreate_test.go [hunk 6]
```diff
@@ -437,10 +379,348 @@ func TestEsdtNFTCreate_ProcessBuiltinFunctionWithExecByCaller(t *testing.T) {
...
+			RecipientAddr: sender.AddressBytes(),
+		}
+
+		vmOutput, err := nftCreate.ProcessBuiltinFunction(sender, nil, vmInput)
+		require.ErrorIs(t, err, ErrInvalidNumberOfArguments)
+		require.True(t, strings.Contains(err.Error(), "for cross chain"))
+		require.Nil(t, vmOutput)
+	})
+
+	t.Run("invalid num of args in exec on dest, penultimate arg is not nonce", func(t *testing.T) {
```
Comment (axenteoctavian) [inline builtInFunctions/esdtNFTCreate_test.go:619]: here we are missing `antepenultimate nonce` and `penultimate creator`

File: builtInFunctions/esdtNFTCreate.go [hunk 7]
```diff
@@ -275,6 +309,64 @@ func getLatestNonce(acnt vmcommon.UserAccountHandler, tokenID []byte) (uint64, e
...
+}
+
+func getCrossChainTokenNonceAndCreator(args [][]byte, callType vm.CallType) (uint64, []byte, error) {
+	minRequiredArgs := minNumOfArgsForCrossChainMint
+	if callType == vm.ExecOnDestByCaller {
+		minRequiredArgs++
+	}
+
+	if len(args) < minRequiredArgs {
+		return 0, nil, fmt.Errorf("%w for cross chain token mint, last 2 arguments should be the nonce and original creator", ErrInvalidNumberOfArguments)
```
Comment (axenteoctavian) [inline builtInFunctions/esdtNFTCreate.go:357]: If the callType is ExecOnDestByCaller, then the comment is not really true...
Maybe add another if at line 353.
Also please print the `len(args)` in the message, might be important some day.

File: builtInFunctions/esdtNFTCreate.go (hunk 5 above)
Comment (mariusmihaic) [reply builtInFunctions/esdtNFTCreate.go:47]: Added it

File: builtInFunctions/esdtNFTCreate_test.go (hunk 6 above)
Comment (mariusmihaic) [reply builtInFunctions/esdtNFTCreate_test.go:619]: Refactored tests

File: builtInFunctions/esdtNFTCreate.go (hunk 7 above)
Comment (mariusmihaic) [reply builtInFunctions/esdtNFTCreate.go]: Refactored

**Review by axenteoctavian — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**